esda/significance.py

```diff
@@ -0,0 +1,164 @@
+
+    Notes
+    -----
+
+    the directed p-value is half of the two-sided p-value, and corresponds to running the
+    lesser and greater tests, then picking the smaller significance value. This is not advised, 
+    since the p-value will be uniformly too small. 
+    """
+    reference_distribution = np.atleast_2d(reference_distribution)
+    n_samples, p_permutations = reference_distribution.shape
+    test_stat = np.atleast_2d(test_stat).reshape(n_samples, -1)
+    if alternative not in (
+        'folded',
+        'two-sided',
+        'greater',
+        'lesser',
+        'directed'
+    ):
+        raise ValueError(
+            f"alternative='{alternative}' provided, but is not"
+            f" one of the supported options: 'two-sided', 'greater', 'lesser', 'directed', 'folded')"
+            )
+    return _permutation_significance(
+        test_stat, 
+        reference_distribution,
+        alternative=alternative
+        )
+
+@njit(parallel=False, fastmath=False)
+def _permutation_significance(test_stat, reference_distribution, alternative='two-sided'):
+    reference_distribution = np.atleast_2d(reference_distribution)
+    n_samples, p_permutations = reference_distribution.shape
+    if alternative == "directed":
+        larger = (reference_distribution >= test_stat).sum(axis=1)
+        low_extreme = (p_permutations - larger) < larger
+        larger[low_extreme] = p_permutations - larger[low_extreme]
+        p_value = (larger + 1.0) / (p_permutations + 1.0)
+    elif alternative == "lesser":
+        p_value = (np.sum(reference_distribution <= test_stat, axis=1) + 1) / (
+            p_permutations + 1
+        )
+    elif alternative == "greater":
+        p_value = (np.sum(reference_distribution >= test_stat, axis=1) + 1) / (
+            p_permutations + 1
+        )
+    elif alternative == "two-sided":
+        # find percentile p at which the test statistic sits
+        # find "synthetic" test statistic at 1-p
+        # count how many observations are outisde of (p, 1-p)
+        # including the test statistic and its synthetic pair
+        percentile = (reference_distribution <= test_stat).mean()*100
+        lows, highs = numpy.percentile(
+            reference_distribution, (percentile, 100-percentile), axis=1
+        ).T
+        n_outside = (reference_distribution <= lows[:,None]).sum(axis=1)
+        n_outside += (reference_distribution >= highs[:,None]).sum(axis=1)
+        p_value = (n_outside + 1) / (p_permutations + 1)
+    elif alternative == "folded":
+        means = reference_distribution.mean(axis=1, keepdims=True)
+        test_stat = np.abs(test_stat - means)
+        reference_distribution = np.abs(reference_distribution - means)
+        p_value = ((reference_distribution >= test_stat).sum(axis=1) + 1) / (
+            p_permutations + 1
+        )
+    else:
+        p_value = np.ones((n_samples, 1))*np.nan
+    return p_value
+
+
+if __name__ == "__main__":
```

Comment:
All below should be turned into tests but I am sure you're aware. Leaving a note to avoid this being merged like this.

Reply:
Yes, I've specified what tests ought to look like now off of this. Will be done w/ benchmarking. We need to prepare both correctness and speed tests for the permutation stuff, since it's so finicky and easy to break.